### PR TITLE
improved rawlogger, blacklisting instead of whitelisting

### PR DESCRIPTION
**Pull request by P-Bent**
The raw logger is now supposed to log every topic except for those which are specifically in an excluded_topics list.
If any topics are listed in the excluded_topics list, the topic will not be logged by the raw logger.

### CODE REVIEW

File: msb/rawlogger/msb_rawlogger.py [hunk 1]
```diff
@@ -15,15 +15,16 @@ def signal_handler(sig, frame):
 
 
 class RawLoggerService:
-    def __init__(self, config: RawLoggerConf, subscriber: ZMQRawSubscriber):
+    def __init__(self, config: RawLoggerConf, subscriber: ZMQRawSubscriber, excluded_topics):
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:18]: The excluded topics are already in the config, it is not necessary to pass them in here again.

File: msb/rawlogger/msb_rawlogger.py [hunk 2]
```diff
@@ -15,15 +15,16 @@ def signal_handler(sig, frame):
...
+        self.excluded_topics = excluded_topics
         self.config = config
         self.subscriber = subscriber
         self.logger = RawLogger(config)
-
+        print(self.logger,"self.logger")
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:23]: What is this print statement for?

File: msb/rawlogger/msb_rawlogger.py [hunk 3]
```diff
@@ -15,15 +15,16 @@ def signal_handler(sig, frame):
...
         if self.config.verbose:
             print(self.config)
 
-    def get_data(self) -> tuple[bytes, bytes]:
+    def get_data(self,excluded_topics) -> tuple[bytes, bytes]:
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:27]: Do not pass `excluded_topics` here, use `self.excluded_topics` if necessary, but it does seem to be used here.

File: msb/rawlogger/msb_rawlogger.py [hunk 4]
```diff
@@ -32,20 +33,25 @@ def get_data(self) -> tuple[bytes, bytes]:
...
             yield topic, data
 
-    def run(self):
-        for topic, data_raw in self.get_data():
-            if self.config.print_stdout:
-                print(f"{topic.decode('utf-8')} : {data_raw.decode('utf-8')}")
 
-            data = b'{"' + topic + b'":' + data_raw + b"}\n"
-            self.logger.write(data)
+    def run(self,excluded_topics):
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:37]: Do not pass `excluded_topics` here, use `self.excluded_topics` instead.

File: msb/rawlogger/msb_rawlogger.py [hunk 5]
```diff
@@ -32,20 +33,25 @@ def get_data(self) -> tuple[bytes, bytes]:
...
+        for topic, data_raw in self.get_data(excluded_topics):
+            if topic.decode('utf-8') in excluded_topics:
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:39]: Decode the topic only once and save it in a variable.

File: msb/rawlogger/msb_rawlogger.py [hunk 6]
```diff
@@ -32,20 +33,25 @@ def get_data(self) -> tuple[bytes, bytes]:
...
+                pass
+            elif self.config.print_stdout:
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:41]: This changes the behavior of the print_stdout flag. This should print additionally to logging, not instead.

File: msb/rawlogger/msb_rawlogger.py (hunk 5 above)
Comment (kawindme) [reply msb/rawlogger/msb_rawlogger.py:39]: Consider using `if topic not  in excluded_topics:` instead. Then you will not need the `pass` and else statements.

File: msb/rawlogger/msb_rawlogger.py [hunk 7]
```diff
@@ -15,15 +15,16 @@ def signal_handler(sig, frame):
 
 
 class RawLoggerService:
-    def __init__(self, config: RawLoggerConf, subscriber: ZMQRawSubscriber):
+    def __init__(self, config: RawLoggerConf, subscriber: ZMQRawSubscriber, excluded_topics):
+        self.excluded_topics = excluded_topics
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:19]: Consider casting `exluding_topics` to a set (`self.excluded_topics = set(excluded_topics)`).
Sets are optimized for frequent "is in" queries.

**Review by kawindme — CHANGES_REQUESTED** [review]
See comments.

But in general please use `black` and `isort` to format files.

File: msb/rawlogger/config.py [hunk 8]
```diff
@@ -1,13 +1,16 @@
...
 
     rollover_period: int = 3600
     """Temporal length in seconds of each logfile: 1h (3600s)"""
 
-    topics: list[str] = ["imu", "gps"]
-    """The topics to log."""
+    excluded_topics: list[str] = ["test"]
+    """The topics to be excluded from log"""
+
+    topics: list[str] = [""]
```
[the remark this answers is not in the document]
Comment (P-Bent) [reply msb/rawlogger/config.py:15]: removed no longer needed list

[the remark this answers is not in the document]
Comment (P-Bent) [reply msb/rawlogger/config.py:15]: removed no longer needed list

File: msb/rawlogger/config.py [hunk 9]
```diff
@@ -1,13 +1,16 @@
...
     data_dir: str = "/home/msb/msb_data/rawlog"
 
     filename_datetime_fmt: str = "%Y%m%dT%H%M%S%z"
 
     rollover_period: int = 3600
     """Temporal length in seconds of each logfile: 1h (3600s)"""
 
-    topics: list[str] = ["imu", "gps"]
-    """The topics to log."""
+    excluded_topics: list[str] = ["test"]
```
[the remark this answers is not in the document]
Comment (P-Bent) [reply msb/rawlogger/config.py]: changed topic "test" to "sta"

File: msb/rawlogger/msb_rawlogger.py (hunk 1 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: Incorperated comment into the code

File: msb/rawlogger/msb_rawlogger.py (hunk 2 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: The print was old and for testing purposes. Print was removed

File: msb/rawlogger/msb_rawlogger.py (hunk 3 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: Removed excluded_topics from get_data method

File: msb/rawlogger/msb_rawlogger.py (hunk 4 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: removed "excluded_topics" as parameter in function call.

File: msb/rawlogger/msb_rawlogger.py (hunk 5 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: Incorperated your suggestion.

File: msb/rawlogger/msb_rawlogger.py (hunk 6 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: New Code version should have fixed this issue

File: msb/rawlogger/msb_rawlogger.py (hunk 7 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: Please give a more detailed explanation of your suggestion.

File: msb/rawlogger/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (P-Bent) [reply msb/rawlogger/config.py:15]: Topics list is needed. [""] subscribes to every zmq topic.

File: msb/rawlogger/msb_rawlogger.py [hunk 10]
```diff
@@ -32,20 +32,24 @@ def get_data(self) -> tuple[bytes, bytes]:
...
-            data = b'{"' + topic + b'":' + data_raw + b"}\n"
-            self.logger.write(data)
 
 
 def main():
     signal.signal(signal.SIGINT, signal_handler)
     raw_logger_config = load_config(RawLoggerConf(), "rawlogger")
-    print(raw_logger_config)
     zmq_config = load_config(ZMQConf(), "zmq")
+    excluded_topics = raw_logger_config.excluded_topics
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:50]: Remove. Not used here.

File: msb/rawlogger/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kawindme) [reply msb/rawlogger/config.py:15]: Fair enough. Please clarify in the docstring that the rawlogger first subscribes to all topics listed in `topics` and that [""] corresponds to all topics. And also clarify that topics in `excluded_topics` are excluded **after** including them in `topics`.
Also consider renaming `topics`  to `included_topics`.

File: msb/rawlogger/msb_rawlogger.py [hunk 11]
```diff
@@ -32,20 +32,24 @@ def get_data(self) -> tuple[bytes, bytes]:
...
             yield topic, data
 
+
     def run(self):
         for topic, data_raw in self.get_data():
-            if self.config.print_stdout:
-                print(f"{topic.decode('utf-8')} : {data_raw.decode('utf-8')}")
+            topic_decoded = topic.decode('utf-8')
+            if topic_decoded not in self.excluded_topics:
+                print(f"{topic_decoded} : {data_raw.decode('utf-8')}")
```
Comment (kawindme) [inline msb/rawlogger/msb_rawlogger.py:40]: Now all data is printed regardless of `self.config.print_stdout`. Please fix.

**Review by kawindme — CHANGES_REQUESTED** [review]

File: msb/rawlogger/config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kawindme) [reply msb/rawlogger/config.py]: I think it was actually a good idea to exclude "test" by default as well.

File: msb/rawlogger/msb_rawlogger.py (hunk 7 above)
Comment (kawindme) [reply msb/rawlogger/msb_rawlogger.py]: Set is another basic data structure in python next to dict and list. It is optimized for frequently querying whether an object is in the set. As you "ask" that question very often in the run method in `if topic_decoded not in self.excluded_topics:` it does make sense to use a set here, i.e. save excluded topics as `self.excluded_topics = set(excluded_topics)` (nothing else has to change). 


(A bit off-topic: Sets in python also support those set operations you know from math, like intersection and difference)

File: msb/rawlogger/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (P-Bent) [reply msb/rawlogger/config.py]: renamed topics to included_topics and added further comments for explanation.

File: msb/rawlogger/config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (P-Bent) [reply msb/rawlogger/config.py]: added "test" topic to .config and .yaml.

File: msb/rawlogger/msb_rawlogger.py (hunk 7 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: Incorporated your suggestion.

File: msb/rawlogger/msb_rawlogger.py (hunk 11 above)
Comment (P-Bent) [reply msb/rawlogger/msb_rawlogger.py]: fixed this issue.

File: config/msb/conf.d/rawlogger.yaml [hunk 12]
```diff
@@ -3,4 +3,4 @@ print_stdout : false
 data_dir : "/home/msb/msb_data/rawlog"
 filename_datetime_fmt : "%Y%m%dT%H%M%S%z"
 rollover_period : 3600  # Temporal length in seconds of each logfile: 1h (3600s)
-topics: ["imu", "gps"]
```
Comment (kawindme) [inline config/msb/conf.d/rawlogger.yaml:6]: `included_topics` is missing in yaml.

File: msb/rawlogger/config.py [hunk 13]
```diff
@@ -1,13 +1,16 @@
...
 
     filename_datetime_fmt: str = "%Y%m%dT%H%M%S%z"
 
     rollover_period: int = 3600
     """Temporal length in seconds of each logfile: 1h (3600s)"""
 
-    topics: list[str] = ["imu", "gps"]
-    """The topics to log."""
+    excluded_topics: list[str] = ["sta","test"]
+    """The topics to be excluded from log. Excluded topics will be removed if "topic" matches element of "excluded_topics", after subscribing to all topics. """
```
Comment (kawindme) [inline msb/rawlogger/config.py:13]: "..., after subscribing to topics listed in `included_topics`"

File: msb/rawlogger/config.py [hunk 14]
```diff
@@ -1,13 +1,16 @@
...
+
+    included_topics: list[str] = [""]
+    """Topic [""] represents all topics. Subscribe to all topics."""
```
Comment (kawindme) [inline msb/rawlogger/config.py:16]: `'Topics to subscribe. Using "" subscribes to all topics.'`

**Review by kawindme — CHANGES_REQUESTED** [review]
Looking good. Just some minor things. 

And please don't forget do use `black` and `isort` to format filed modified.

File: msb/rawlogger/config.py (hunk 13 above)
Comment (P-Bent) [reply msb/rawlogger/config.py]: Done

File: msb/rawlogger/config.py (hunk 14 above)
Comment (P-Bent) [reply msb/rawlogger/config.py]: Done

File: config/msb/conf.d/rawlogger.yaml (hunk 12 above)
Comment (P-Bent) [reply config/msb/conf.d/rawlogger.yaml:6]: Done

**Review by kawindme — APPROVED** [review]
Nice work, thanks.

**Pull request merged**